windows/hid.c

```diff
@@ -1146,13 +1369,1019 @@ int HID_API_EXPORT_CALL HID_API_CALL hid_get_indexed_string(hid_device *dev, int
+
+				if (rt_idx == HidP_Input) {
+					// Write "Input" main item
+					rd_write_short_item(rd_main_input, pp_data->caps[caps_idx].BitField, &byte_list);
+				}
+				else if (rt_idx == HidP_Output) {
+					// Write "Output" main item
+					rd_write_short_item(rd_main_output, pp_data->caps[caps_idx].BitField, &byte_list);
+				}
+				else if (rt_idx == HidP_Feature) {
+					// Write "Feature" main item
+					rd_write_short_item(rd_main_feature, pp_data->caps[caps_idx].BitField, &byte_list);
+				}
+				report_count = 0;
+			}
+		}
+        // Go to next item in main_item_list and free the memory of the actual item 
+        struct rd_main_item_node* main_item_list_prev = main_item_list;
+		main_item_list = main_item_list->next;
+        free(main_item_list_prev);        
+	}
+
+	// Free multidimensionable array: coll_bit_range[COLLECTION_INDEX][REPORT_ID][INPUT/OUTPUT/FEATURE]
+	// Free multidimensionable array: coll_child_order[COLLECTION_INDEX][DIRECT_CHILD_INDEX]
+	for (USHORT collection_node_idx = 0; collection_node_idx < pp_data->NumberLinkCollectionNodes; collection_node_idx++) {
+		for (int reportid_idx = 0; reportid_idx < 256; reportid_idx++) {
+			for (HIDP_REPORT_TYPE rt_idx = 0; rt_idx < NUM_OF_HIDP_REPORT_TYPES; rt_idx++) {
+				free(coll_bit_range[collection_node_idx][reportid_idx][rt_idx]);
+			}
+			free(coll_bit_range[collection_node_idx][reportid_idx]);
+		}
+		free(coll_bit_range[collection_node_idx]);
+		if (coll_number_of_direct_childs[collection_node_idx] != 0) free(coll_child_order[collection_node_idx]);
+	}
+	free(coll_bit_range);
+	free(coll_child_order);
+
+	// Free one dimensional arrays
+	free(coll_begin_lookup);
+	free(coll_end_lookup);
+    free(coll_levels);
+    free(coll_number_of_direct_childs);
+	    
+	// Copy report temporary descriptor list into buf array and free list
+	unsigned int byte_list_len = 0;
+	while ((byte_list != NULL))
+	{
+		if (byte_list_len < buf_size) {
+			// In case of too small buf size, just inhibt write to buffer,
+			// to ensure proper free of list memory
+			*(buf + byte_list_len) = (unsigned char)byte_list->byte;
+		}
+		byte_list_len++;
+		struct rd_item_byte* byte_list_prev = byte_list;
+		byte_list = byte_list->next;
+		free(byte_list_prev);
+	}
+
+	if (byte_list_len > buf_size) {
+		register_error_msg(dev, L"Buffer to small for the generated report descriptor");
+		return -1;
+	}
+	else {
+		return byte_list_len;
+	}
+}
+
+int HID_API_EXPORT_CALL hid_get_report_descriptor(hid_device* dev, unsigned char* buf, size_t buf_size)
 {
-	(void)dev;
-	(void)buf;
-	(void)buf_size;
-	/* TODO  */
-	return -1;
+	PHIDP_PREPARSED_DATA pp_data = NULL;
+
+	if (!HidD_GetPreparsedData(dev->device_handle, &pp_data)) {
+		register_error(dev, "HidD_GetPreparsedData");
+		return -1;
+	}
+	
+	int res;
+	res = rd_reconstructor(dev, pp_data, buf, buf_size);
+
+	if (!HidD_FreePreparsedData(pp_data)) {
+		register_error(dev, "HidD_FreePreparsedData");
+		return -1;
+	}
```

Comment:
this cannot fail, unless memory corruption, in which case we're already "screwed"
```suggestion
	(void)HidD_FreePreparsedData(pp_data);
```